src/Terminal/commands/grep.ts

```diff
@@ -0,0 +1,179 @@
+}
+
+function parseFile(parseFunc: PtLineParser): FileParser {
+  return function (file: ContentFile | null): string | string[] {
+    if (!file) return "";
+    const content: string = "content" in file ? file["content"] : file["code"];
+
+    const parseLine: LineParser = parseFunc.bind(null, file.filename);
+
+    const editedContent: string | string[] = content.split("\n").map(parseLine);
+
+    return editedContent;
+  };
+}
+
+function getServerFiles(server: BaseServer): [ContentFile[], string[]] {
+  return [
+    [...allContentFiles(server)].map((tuple: FileTuple): ContentFile => tuple[1]),
+    [], // empty array for badFiles
+  ];
+}
+
+function getArgFiles(args: string[]): [(ContentFile | null)[], string[]] {
+  const notFiles: string[] = [];
+  const files: (ContentFile | null)[] = args.map((arg: string): ContentFile | null => {
+    const script: ContentFile | null = hasTextExtension(arg) ? Terminal.getTextFile(arg) : Terminal.getScript(arg);
+    if (!script) {
+      notFiles.push(arg);
+    }
+    return script;
+  });
+  return [files, notFiles];
+}
+
+function filterOpts([options, otherArgs]: [Options, string[]], arg: string): [Options, string[]] {
+  const isOption: boolean = Object.keys(OK_ARGS).some((key: string): boolean =>
+    OK_ARGS[key as keyof OkArgs].includes(arg) ? (options[key as keyof Options] = true) : false,
+  );
+  return isOption ? [options, otherArgs] : [options, [...otherArgs, arg]];
+}
+
+function filterArgs(args: string[]): [Options, string[], FilePath | null] {
+  const initOpts: Options = {
+    lineNum: false,
+    regExpr: false,
+    yesName: false,
+    notName: false,
+    multiFile: false,
+    toFile: false,
+    overWrt: false,
+    quiet: false,
+  };
+
+  const outputArgIndex = OK_ARGS.toFile.reduce((ret: number, arg: string) => {
+    const argIndex = args.indexOf(arg);
+    return argIndex > -1 ? argIndex : ret;
+  }, NaN);
+  // if outputArgIndex !NaN grab next arg as output file
+  const outFileStr: string | undefined = !isNaN(outputArgIndex) ? args.splice(outputArgIndex + 1)?.[0] : undefined;
+
+  const outFile = outFileStr ? Terminal.getFilepath(outFileStr) : null;
+
+  const [options, otherArgs]: [Options, string[]] = args.map(String).reduce(filterOpts, [initOpts, []]);
+
+  options.toFile = !!outFile;
+
+  return [options, otherArgs, outFile];
+}
+
+export function grep(args: (string | number | boolean)[], server: BaseServer): void {
+  if (!args.length) {
+    return Terminal.error("Incorrect usage of grep command. Usage: grep [OPTION]... PATTERN [FILE]...");
+  }
+
+  const [options, otherArgs, outFilePath]: [Options, string[], FilePath | null] = filterArgs(args.map(String));
+  if (!outFilePath || !hasTextExtension(outFilePath)) {
+    return Terminal.error(`grep failed: Invalid output file "${outFilePath}". Output file must be a text file.`);
+  }
+
+  console.log(options);
+  if (options.toFile && !options.overWrt) {
+    const overWrtFileExists = [...allContentFiles(server)]
```

Comment:
Optional: This one makes me so itchy I had to leave a comment XD
This makes two extra copies of the data, when instead you could do this in a single pass with no copies:
```suggestion
    for (const tuple of allContentFiles(server)) {
      if (tuple[1].filename === outFilePath) {
        return Terminal.error(
          `grep failed: Invalid output file "${outFilePath}". Output file must not already exist. Pass -f to force.`,
        );
      }
    }
```

Reply:
Ooh neat. I wasn't aware that for of could iterate through that without arrayifying, like the dot methods need. :+1: